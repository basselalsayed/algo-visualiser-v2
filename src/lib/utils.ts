import { type ClassValue, clsx } from 'clsx';
import { twMerge } from 'tailwind-merge';

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

type T0 =
  | 'string'
  | 'number'
  | 'bigint'
  | 'boolean'
  | 'symbol'
  | 'undefined'
  | 'object'
  | 'function';

export function assert(value: unknown, type: T0): asserts value is T0 {
  const actualType = typeof value;

  if (type === 'object') {
    if (value === null) {
      throw new TypeError(`Expected type '${type}', but received null`);
    }
  }

  if (actualType !== type) {
    throw new TypeError(
      `Expected type '${type}', but received '${actualType}'`
    );
  }
}

export function typeIs(value: unknown, type: T0): value is T0 {
  try {
    assert(value, type);
    return true;
  } catch {
    return false;
  }
}

export function convertToMilliseconds(
  start: DOMHighResTimeStamp,
  end: DOMHighResTimeStamp
) {
  const milliseconds = end - start;
  return Math.floor(milliseconds * 1000) / 1000; // 3 decimal places
}

export function convertToSeconds(
  start: DOMHighResTimeStamp,
  end: DOMHighResTimeStamp
) {
  const seconds = (end - start) / 1000;
  return Math.floor(seconds * 1000) / 1000;
}

export function elementIsFullyScrolled(element: HTMLElement): boolean {
  return (
    Math.floor(element.scrollHeight - element.scrollTop) ===
    Math.floor(element.clientHeight)
  );
}
