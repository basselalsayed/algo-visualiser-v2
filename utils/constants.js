import path from 'node:path';

export const DIRECTORIES = [
  'algorithms',
  'components',
  'contexts',
  'data',
  'hooks',
  'lib',
  'store',
];

export const ROOT_PATH = path.resolve(import.meta.dirname, '..');
