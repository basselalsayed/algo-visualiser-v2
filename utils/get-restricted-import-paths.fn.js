import fs from 'node:fs';
import path from 'node:path';

const __dirname = new URL('.', import.meta.url).pathname;

/**
 * Recursively walks through a directory and collects relative paths to all index.ts or index.tsx files.
 *
 * @param {string} dir - Absolute path to start walking from.
 * @param {string} baseRelativePath - Relative path from the module root (used for nested structure).
 * @returns {string[]} - List of relative paths to directories containing index files.
 */
function findIndexFileDirs(dir, baseRelativePath = '') {
  const indexDirs = [];
  const entries = fs.readdirSync(dir);

  for (const entry of entries) {
    const fullPath = path.join(dir, entry);
    const relPath = path.join(baseRelativePath, entry);
    const stat = fs.statSync(fullPath);

    if (stat.isDirectory()) {
      indexDirs.push(...findIndexFileDirs(fullPath, relPath));
    } else if (
      (entry === 'index.ts' || entry === 'index.tsx') &&
      baseRelativePath !== ''
    ) {
      indexDirs.push(baseRelativePath);
    }
  }

  return indexDirs;
}

/**
 * Generates a list of restricted import paths for all index.ts(x) files in a module.
 *
 * @param {string} dirAbsolutePath - Absolute path to the module directory (e.g. src/components)
 * @param {string} alias - Import alias for the module (e.g. @/components)
 * @returns {Array<{ name: string, message: string }>}
 */
function getRestrictedIndexImports(dirAbsolutePath, alias) {
  const indexDirs = findIndexFileDirs(dirAbsolutePath);

  const staticUpwardPaths = [
    '../../../index',
    '../../..',
    '../../index',
    '../..',
    '../index',
    '..',
  ].map((name) => ({
    message: `Avoid importing '${name}'. Use aliased or direct file paths.`,
    name,
  }));

  const indexPathViolations = indexDirs.flatMap((relPath) => [
    {
      message: `Avoid importing index file: ${alias}/${relPath}/index.ts. Import specific files instead.`,
      name: `${alias}/${relPath}`,
    },
    {
      message: `Avoid importing index file: ../${relPath}/index.ts. Import specific files instead.`,
      name: `../${relPath}`,
    },
  ]);

  return [...staticUpwardPaths, ...indexPathViolations];
}

export { __dirname, getRestrictedIndexImports };
