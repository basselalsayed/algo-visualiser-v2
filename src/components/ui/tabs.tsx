import * as TabsPrimitive from '@radix-ui/react-tabs';
import * as React from 'react';

import { cn } from '@/lib';

const Tabs = TabsPrimitive.Root;

const TabsList = ({
  className,
  ref,
  ...props
}: React.ComponentPropsWithRef<typeof TabsPrimitive.List>) => (
  <TabsPrimitive.List
    ref={ref}
    className={cn(
      'inline-flex items-center justify-center rounded-lg bg-muted p-1 text-muted-foreground',
      className
    )}
    {...props}
  />
);
TabsList.displayName = TabsPrimitive.List.displayName;

const TabsTrigger = ({
  className,
  ref,
  ...props
}: React.ComponentPropsWithRef<typeof TabsPrimitive.Trigger>) => (
  <TabsPrimitive.Trigger
    ref={ref}
    className={cn(
      'inline-flex items-center justify-center rounded-md px-3 py-1 text-sm font-medium whitespace-nowrap ring-offset-background transition-all focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 focus-visible:outline-hidden disabled:pointer-events-none disabled:opacity-50 data-[state=active]:bg-background data-[state=active]:text-foreground data-[state=active]:shadow-sm',
      className
    )}
    {...props}
  />
);
TabsTrigger.displayName = TabsPrimitive.Trigger.displayName;

const TabsContent = ({
  className,
  ref,
  ...props
}: React.ComponentPropsWithRef<typeof TabsPrimitive.Content>) => (
  <TabsPrimitive.Content
    ref={ref}
    className={cn(
      'mt-2 ring-offset-background focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 focus-visible:outline-hidden',
      className
    )}
    {...props}
  />
);
TabsContent.displayName = TabsPrimitive.Content.displayName;

export { Tabs, TabsContent, TabsList, TabsTrigger };
