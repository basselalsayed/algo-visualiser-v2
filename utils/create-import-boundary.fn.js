import path from 'node:path';

import { DIRECTORIES, ROOT_PATH } from './constants.js';
import { getRestrictedIndexImports } from './get-restricted-import-paths.fn.js';

/**
 * @param {string} dir
 */
const getAlias = (dir) => `@/${dir}`;

/**
 * @param {string} dir
 */
export function createImportBoundary(dir) {
  const alias = getAlias(dir);
  const glob = `src/${dir}/**/*.{ts,tsx}`;

  const forbid = DIRECTORIES.flatMap((_dir) =>
    _dir === dir ? undefined : [`${getAlias(_dir)}/**/*`, `${getAlias(_dir)}/*`]
  ).filter(Boolean);

  const absoluteDir = path.resolve(ROOT_PATH, 'src', dir);

  const restrictedIndexPaths = getRestrictedIndexImports(absoluteDir, alias);

  return {
    files: [glob],
    rules: {
      'import/no-internal-modules': [
        'error',
        {
          forbid,
        },
      ],
      'no-restricted-imports': [
        'error',
        {
          paths: [
            ...restrictedIndexPaths,
            {
              message: `Avoid importing '${alias}' from within ${dir}/. Use direct file paths.`,
              name: alias,
            },
          ],
          patterns: [
            {
              group: ['**/index', '**/index.ts', '**/index.tsx'],
              message: `Avoid importing index files directly. Import specific files instead.`,
            },
            {
              group: ['../../**/*'],
              message: `Avoid relative imports outside of this module. Use alias paths instead.`,
            },
          ],
        },
      ],
    },
  };
}
