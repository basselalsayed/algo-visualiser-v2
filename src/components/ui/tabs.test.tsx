import { renderFragmentSnapshot } from '@/test/utils';

import { Tabs, TabsList, TabsTrigger } from './tabs';

describe('Tabs.component', () => {
  it('looks the same', () => {
    expect(
      renderFragmentSnapshot(
        <Tabs value='1'>
          <TabsList>
            <TabsTrigger value='1'>1</TabsTrigger>
            <TabsTrigger value='2'>2</TabsTrigger>
          </TabsList>
        </Tabs>
      )
    ).toMatchSnapshot();
    expect(
      renderFragmentSnapshot(
        <Tabs>
          <TabsList>
            <TabsTrigger value='1'>1</TabsTrigger>
            <TabsTrigger value='2'>2</TabsTrigger>
          </TabsList>
        </Tabs>
      )
    ).toMatchSnapshot();
  });
});
