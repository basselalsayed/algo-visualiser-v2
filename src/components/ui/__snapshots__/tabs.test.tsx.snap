// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tabs.component > looks the same 1`] = `
<DocumentFragment>
  <div
    data-orientation="horizontal"
    dir="ltr"
  >
    <div
      aria-orientation="horizontal"
      class="inline-flex items-center justify-center rounded-lg bg-muted p-1 text-muted-foreground"
      data-orientation="horizontal"
      role="tablist"
      style="outline: none;"
      tabindex="0"
    >
      <button
        aria-controls="radix-«r0»-content-1"
        aria-selected="true"
        class="inline-flex items-center justify-center rounded-md px-3 py-1 text-sm font-medium whitespace-nowrap ring-offset-background transition-all focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 focus-visible:outline-hidden disabled:pointer-events-none disabled:opacity-50 data-[state=active]:bg-background data-[state=active]:text-foreground data-[state=active]:shadow-sm"
        data-orientation="horizontal"
        data-radix-collection-item=""
        data-state="active"
        id="radix-«r0»-trigger-1"
        role="tab"
        tabindex="-1"
        type="button"
      >
        1
      </button>
      <button
        aria-controls="radix-«r0»-content-2"
        aria-selected="false"
        class="inline-flex items-center justify-center rounded-md px-3 py-1 text-sm font-medium whitespace-nowrap ring-offset-background transition-all focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 focus-visible:outline-hidden disabled:pointer-events-none disabled:opacity-50 data-[state=active]:bg-background data-[state=active]:text-foreground data-[state=active]:shadow-sm"
        data-orientation="horizontal"
        data-radix-collection-item=""
        data-state="inactive"
        id="radix-«r0»-trigger-2"
        role="tab"
        tabindex="-1"
        type="button"
      >
        2
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tabs.component > looks the same 2`] = `
<DocumentFragment>
  <div
    data-orientation="horizontal"
    dir="ltr"
  >
    <div
      aria-orientation="horizontal"
      class="inline-flex items-center justify-center rounded-lg bg-muted p-1 text-muted-foreground"
      data-orientation="horizontal"
      role="tablist"
      style="outline: none;"
      tabindex="0"
    >
      <button
        aria-controls="radix-«r3»-content-1"
        aria-selected="false"
        class="inline-flex items-center justify-center rounded-md px-3 py-1 text-sm font-medium whitespace-nowrap ring-offset-background transition-all focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 focus-visible:outline-hidden disabled:pointer-events-none disabled:opacity-50 data-[state=active]:bg-background data-[state=active]:text-foreground data-[state=active]:shadow-sm"
        data-orientation="horizontal"
        data-radix-collection-item=""
        data-state="inactive"
        id="radix-«r3»-trigger-1"
        role="tab"
        tabindex="-1"
        type="button"
      >
        1
      </button>
      <button
        aria-controls="radix-«r3»-content-2"
        aria-selected="false"
        class="inline-flex items-center justify-center rounded-md px-3 py-1 text-sm font-medium whitespace-nowrap ring-offset-background transition-all focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 focus-visible:outline-hidden disabled:pointer-events-none disabled:opacity-50 data-[state=active]:bg-background data-[state=active]:text-foreground data-[state=active]:shadow-sm"
        data-orientation="horizontal"
        data-radix-collection-item=""
        data-state="inactive"
        id="radix-«r3»-trigger-2"
        role="tab"
        tabindex="-1"
        type="button"
      >
        2
      </button>
    </div>
  </div>
</DocumentFragment>
`;
